test: fix an incorrect `feature_fee_estimation.py` subtest

ismaelsadeeq opened the pull request:
Attempt to fix #32461 and #31944

In the `estimatesmartfee` RPC, we return the maximum of the following: the feerate estimate for the target, `minrelaytxfee`, and `mempoolminfee`.

https://github.com/bitcoin/bitcoin/blob/9a05b45da60d214cb1e5a50c3d2293b1defc9bb0/src/rpc/fees.cpp#L85

The test `test_feerate_mempoolminfee`, originally introduced in https://github.com/bitcoin/bitcoin/commit/ea31caf6b4c182c6f10f136548f84e603800511c, is incorrect. 

It should calculate the fee rate ceiling by taking the maximum of the custom `minrelaytxfee`, `mempoolminfee`, and the highest fee rate observed during the test (`check_smart_estimates`). This is necessary because:

* There is no guarantee that the generated fee rates will exceed both `minrelaytxfee` and `mempoolminfee`.
* Users can start a node with custom fee settings.

Due to the non-deterministic nature of the `feature_fee_estimation.py` test, it often passes by chance. The randomly generated fees typically include a value higher than the custom `minrelaytxfee`, inadvertently hiding the issue.

Issue #32461 identified a random seeds that consistently fails the test because the generated fees never exceed the custom `minrelaytxfee`:

e.g
```
build/test/functional/feature_fee_estimation.py --random=3450808900320758527
```

This PR has two commits which :

* Correctly fixes the test by calculating the fee rate ceiling as the maximum of the node `minrelaytxfee`, `mempoolminfee`, and the highest seen fee rate, when verifying smart fee estimates.
* Improves the subtest name and comment for clarity.
* Restores the original test behavior by appending 4000 WU to the custom `blockmaxweight`.


It also fixes #31944

Before this PR we shrank the blocks by 4000, reducing the number of txs that are mined, which effectively causes the long stats to not have insufficient data.
See https://github.com/bitcoin/bitcoin/issues/31944#issuecomment-2699259294

Increasing the weight by 4000 (matching the original test block weight) allows for more transactions to be mined and the long stats to have enough data.  

credits to maflcko for noticing the test no longer fails after this https://github.com/bitcoin/bitcoin/issues/31944#issuecomment-3982398137

l0rinc commented:
Thanks for fixing it so quickly, I can confirm that this fixes the test for the failing `--random=3450808900320758527`.
Someone more knowledgeable in this area of the code should also look at at, but from my part it's a lightweight ACK.

ACK 2c5f26bc6ac47e3a6a8555e9a6c60832322a36e8

willcl-ark reviewed (approved):
ACK 2c5f26bc6ac47e3a6a8555e9a6c60832322a36e8

If you re-touch for any reason it would seem a little more logical to me to slightly restructure so we test without new `minrelaytxfee` first (`check_estimates()` on L260), then add the high `minrelaytxfee` before doing the rest of the tests with it added. But zero practical difference and not a blocker for me.

ismaelsadeeq commented:
cc @glozow Is this rfm? I'd like to make a quick follow-up on this, as mentioned in 
 https://github.com/bitcoin/bitcoin/pull/32463#issuecomment-2930709631

Also, this can be backported to 29.1 cc @fanquake.

glozow reviewed (commented):
Nice work finding the issue. For some reason, I couldn't reproduce using `--random=3450808900320758527`. But `--random=4130785039185616282` from https://cirrus-ci.com/task/5487636053229568 and `--random=6565824249277094228` from https://github.com/bitcoin/bitcoin/runs/43526123006 worked for me.

Did you consider baking this logic into `check_smart_estimates` instead? It feels a bit hacky to just append it to the fees seen, given that it isn't actually a seen fee.

```
diff --git a/test/functional/feature_fee_estimation.py b/test/functional/feature_fee_estimation.py
index 61db186de17..07dabc6e40f 100755
--- a/test/functional/feature_fee_estimation.py
+++ b/test/functional/feature_fee_estimation.py
@@ -91,19 +91,21 @@ def check_smart_estimates(node, fees_seen):
     """Call estimatesmartfee and verify that the estimates meet certain invariants."""
 
     delta = 1.0e-6  # account for rounding error
-    last_feerate = float(max(fees_seen))
     all_smart_estimates = [node.estimatesmartfee(i) for i in range(1, 26)]
     mempoolMinFee = node.getmempoolinfo()["mempoolminfee"]
     minRelaytxFee = node.getmempoolinfo()["minrelaytxfee"]
+    feerate_ceiling = max(max(fees_seen), float(mempoolMinFee), float(minRelaytxFee))
+
+    last_feerate = feerate_ceiling
     for i, e in enumerate(all_smart_estimates):  # estimate is for i+1
         feerate = float(e["feerate"])
         assert_greater_than(feerate, 0)
         assert_greater_than_or_equal(feerate, float(mempoolMinFee))
         assert_greater_than_or_equal(feerate, float(minRelaytxFee))
 
-        if feerate + delta < min(fees_seen) or feerate - delta > max(fees_seen):
+        if feerate + delta < min(fees_seen) or feerate - delta > feerate_ceiling:
             raise AssertionError(
-                f"Estimated fee ({feerate}) out of range ({min(fees_seen)},{max(fees_seen)})"
+                f"Estimated fee ({feerate}) out of range ({min(fees_seen)},{feerate_ceiling})"
             )
         if feerate - delta > last_feerate:
             raise AssertionError(
@@ -238,12 +240,13 @@ class EstimateFeeTest(BitcoinTestFramework):
         self.log.info("Final estimates after emptying mempools")
         check_estimates(self.nodes[1], self.fees_per_kb)
 
-    def test_feerate_mempoolminfee(self):
+    def test_estimates_with_highminrelaytxfee(self):
         high_val = 3 * self.nodes[1].estimatesmartfee(1)["feerate"]
         self.restart_node(1, extra_args=[f"-minrelaytxfee={high_val}"])
         check_estimates(self.nodes[1], self.fees_per_kb)
         self.restart_node(1)
 
+
     def sanity_check_rbf_estimates(self, utxos):
         """During 5 blocks, broadcast low fee transactions. Only 10% of them get
         confirmed and the remaining ones get RBF'd with a high fee transaction at
@@ -452,11 +455,11 @@ class EstimateFeeTest(BitcoinTestFramework):
         self.log.info("Test fee_estimates.dat is flushed periodically")
         self.test_estimate_dat_is_flushed_periodically()
 
-        # check that the effective feerate is greater than or equal to the mempoolminfee even for high mempoolminfee
+        # check that estimatesmartfee feerate is greater than or equal to maximum of mempoolminfee and minrelaytxfee
         self.log.info(
-            "Test fee rate estimation after restarting node with high MempoolMinFee"
+            "Test fee rate estimation after restarting node with high minrelaytxfee"
         )
-        self.test_feerate_mempoolminfee()
+        self.test_estimates_with_highminrelaytxfee()
 
         self.log.info("Test acceptstalefeeestimates option")
         self.test_acceptstalefeeestimates_option()
```

ismaelsadeeq committed:
test: fix incorrect subtest in `feature_fee_estimation.py`

- Update `check_smart_estimates` to calculate the fee rate ceiling
   by taking the maximum of fees seen, minrelaytxfee, and mempoolminfee.
- Improve the subtest name and comments.

ismaelsadeeq committed:
test: retain the intended behavior of `feature_fee_estimation.py` nodes

- Increase block weight by 4000 for all nodes with custom -blockmaxweight.
  Prior to this commit, we generated blocks with 4000 weight units less worth of transactions.
  See https://github.com/bitcoin/bitcoin/issues/32461#issuecomment-2925282272 for details.
  This commit fixes it by increasing the block weight by 4000.

ismaelsadeeq commented:
Thanks for you review @glozow 

> Did you consider baking this logic into check_smart_estimates instead? It feels a bit hacky to just append it to the fees seen, given that it isn't actually a seen fee.

I hadn’t considered that before, your suggestion is definitely more elegant and cleaner.

I’ve updated the PR to implement your approach, and I also added a follow-up improvement I had in mind as a separate commit.

[22a36e8..9b75cfd](https://github.com/bitcoin/bitcoin/compare/2c5f26bc6ac47e3a6a8555e9a6c60832322a36e8..9b75cfda4d62a0a3bde402503244dd57e1621a12)

glozow reviewed (commented):
ACK 9b75cfda4d62a0a3bde402503244dd57e1621a12

theStack reviewed (approved):
Light ACK 9b75cfda4d62a0a3bde402503244dd57e1621a12

I'm not familiar with the inner workings of our fee estimation code, but only considering how the "feerate" result is constructed in the `estimatesmartfee` RPC, the changes look good to me. Increasing the max block weight params by 4000 to compensate for https://github.com/bitcoin/bitcoin/pull/31384 also makes sense.

> Nice work finding the issue. For some reason, I couldn't reproduce using `--random=3450808900320758527`. But `--random=4130785039185616282` from https://cirrus-ci.com/task/5487636053229568 and `--random=6565824249277094228` from https://github.com/bitcoin/bitcoin/runs/43526123006 worked for me.

Same here, fwiw, running with `--random=3450808900320758527` didn't fail for me on master (at commit bf75c9964fb28a5e9b8a07097248ffa960738478).

achow101 commented:
ACK 9b75cfda4d62a0a3bde402503244dd57e1621a12

fanquake commented:
> Same here, fwiw, running with --random=3450808900320758527 didn't fail for me on master (at commit https://github.com/bitcoin/bitcoin/commit/bf75c9964fb28a5e9b8a07097248ffa960738478).

This is the one that fails for me on 29.x. Backported to 29.x in #32863.
